[view-transitions] Create renderers for pseudo-elements

nt1m opened the pull request:
#### 5d7c3bb3e28035d21e4461474204eac7493419c4
<pre>
[view-transitions] Create renderers for pseudo-elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=265168">https://bugs.webkit.org/show_bug.cgi?id=265168</a>
<a href="https://rdar.apple.com/118667022">rdar://118667022</a>

Reviewed by Antti Koivisto.

Use the named elements map as input to build the pseudo element tree.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::isEmpty const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::viewTransitionRoot const):
(WebCore::RenderView::setViewTransitionRoot):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::functionalPseudoElementArgument const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setFunctionalPseudoElementArgument):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::RenderTreeUpdater):
(WebCore::RenderTreeUpdater::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeUpdater.h:
(WebCore::RenderTreeUpdater::viewTransition):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp: Added.
(WebCore::RenderTreeUpdater::ViewTransition::ViewTransition):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h: Added.
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):

Canonical link: <a href="https://commits.webkit.org/273336@main">https://commits.webkit.org/273336@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ba26722a4fb260593199570642213e6c130e601

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14006 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37845 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/31684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36246 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11076 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35630 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11862 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31299 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10393 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39092 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31728 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8497 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34453 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8043 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11093 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/11412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdater.h:
```diff
@@ -84,6 +85,7 @@ class RenderTreeUpdater {
     RenderTreePosition& renderTreePosition();
 
     GeneratedContent& generatedContent() { return *m_generatedContent; }
+    ViewTransition& viewTransition() { return *m_viewTransition; }
```
Should we check/assert about null pointer here ? (Should we use a pointer in the first place?)

nt1m replied:
I just followed the same pattern as GeneratedContent

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -0,0 +1,238 @@
+    Vector<SingleThreadWeakPtr<RenderObject>> descendantsToDelete;
+    auto* currentGroup = documentElementRenderer.view().viewTransitionRoot()->firstChild();
+    for (auto& name : activeViewTransition->namedElements().keys()) {
+        ALWAYS_LOG_WITH_STREAM(stream << "building " << name);
```
forgotten log

nt1m replied:
will remove

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -0,0 +1,238 @@
+    // Traverse named elements map to update/build all ::view-transition-group().
+    Vector<SingleThreadWeakPtr<RenderObject>> descendantsToDelete;
+    auto* currentGroup = documentElementRenderer.view().viewTransitionRoot()->firstChild();
+    for (auto& name : activeViewTransition->namedElements().keys()) {
```
`const auto&`?

nt1m replied:
I don't think it makes a huge difference, but I could use it

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -0,0 +1,238 @@
+
+    // Create / remove ::view-transtion-image-pair itself.
+    SingleThreadWeakPtr<RenderElement> imagePair;
+    if ((imagePair = downcast<RenderElement>(group.firstChild()))) {
```
This looks weird. Should it be a dynamicDowncast ?

nt1m replied:
Not really, this is null checking `group.firstChild()`. If `group.firstChild()` is present, it's guaranteed to be a `RenderElement`. `uncheckedDowncast` could work here.

mdubet replied:
Maybe we should write the simpler `if (group.firstChild())` and do the downcast in the block ?

annevk commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Nit: Happy new year Tim!

anttijk commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -670,6 +670,7 @@ inline bool RenderStyle::specifiesColumns() const { return !hasAutoColumnCount()
 constexpr OptionSet<Containment> RenderStyle::strictContainment() { return { Containment::Size, Containment::Layout, Containment::Paint, Containment::Style }; }
 inline StyleColor RenderStyle::strokeColor() const { return m_rareInheritedData->strokeColor; }
 inline float RenderStyle::strokeMiterLimit() const { return m_rareInheritedData->miterLimit; }
+inline const AtomString& RenderStyle::styleTypeIdentifier() const { return m_nonInheritedData->rareData->styleTypeIdentifier; }
```
This should be called something more informative, either something specific to this feature like `viewTransitionPseudoElementNameSelector` or something more generic like `functionalPseudoElementArgument`. Of course `styleType()` is weird in itself, it should be `pseudoElementType()` or similar.

nt1m replied:
will change to `functionalPseudoElementArgument` since `::highlight()` gets this as well. I'll rename `styleType()` in a separate patch.

anttijk commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -285,6 +285,7 @@ inline void RenderStyle::setSpecifiedZIndex(int value) { SET_NESTED_PAIR(m_nonIn
 inline void RenderStyle::setStrokeColor(const StyleColor& color) { SET(m_rareInheritedData, strokeColor, color); }
 inline void RenderStyle::setStrokeMiterLimit(float value) { SET(m_rareInheritedData, miterLimit, value); }
 inline void RenderStyle::setStrokeWidth(Length&& width) { SET(m_rareInheritedData, strokeWidth, WTFMove(width)); }
+inline void RenderStyle::setStyleTypeIdentifier(const AtomString& identifier) { SET_NESTED(m_nonInheritedData, rareData, styleTypeIdentifier, identifier); }
```
It would be good to assert this does not get set except for right pseudo-elements.

nt1m replied:
will do

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -0,0 +1,237 @@
+
+// The contents and ordering of the named elements map should remain stable during the duration of the transition.
+// We should only need to handle changes in the `display` CSS property by recreating / deleting renderers as needed.
+void RenderTreeUpdater::ViewTransition::updatePseudoElementTree(RenderElement& documentElementRenderer)
```
Is there a good test coverage for mutations? You should ensure these pseudo elements are added to the fuzzers soon.

nt1m replied:
WPTs cover mutations to some extent, but right now proper rendering test coverage can't be done until we hook this up to animation timing because pseudo-elements won't appear for long enough (this is what I'm planning to look into next).

nt1m committed:
[view-transitions] Create renderers for pseudo-elements
https://bugs.webkit.org/show_bug.cgi?id=265168
rdar://118667022

Reviewed by Antti Koivisto.

Use the named elements map as input to build the pseudo element tree.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::isEmpty const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::viewTransitionRoot const):
(WebCore::RenderView::setViewTransitionRoot):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::functionalPseudoElementArgument const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setFunctionalPseudoElementArgument):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::RenderTreeUpdater):
(WebCore::RenderTreeUpdater::updateAfterDescendants):
* Source/WebCore/rendering/updating/RenderTreeUpdater.h:
(WebCore::RenderTreeUpdater::viewTransition):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp: Added.
(WebCore::RenderTreeUpdater::ViewTransition::ViewTransition):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h: Added.
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):

Canonical link: https://commits.webkit.org/273336@main
